docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+Open Enclave Protected Code Loader
+====
+
+Enabling Enclave Code Confidentiality in Open Enclave.
+
+# Motivation
+Current Open Enclave SDK provides code integrity and data confidentiality at
+run-time but not enclave binary confidentiality on a disk. The enclave binary
+is in plaintext and can be reverse engineered to reveal code logic and
+secret data embedded in the enclave image. 
+
+The Open Enclave Protected Code Loader shall enable providing confidentiality 
+and integrity to the IP sections in enclave image, i.e. , the ELF file.
+
+# User Experience
+## Enclave Build-time:
+### Linking pcl lib to the enclave 
+On building the enclave, a static lib "liboepcl.a" should be linked to the 
+enclave image with the compilation option "-Wl, whole-archive"
+### Encrypt the enclave image
+An encryption tool is provided as a pre-step to the enclave signing process:
+
+```bash
+oeencrypt -i my_enclave -o my_enclave.enc -k keyfile
+```
+
+## Enclave Load-time:
+On Enclave Loading, if the image is encrypted, it needs to be decrypted 
+inside the enclave loader, this step is done by the OE SDK runtime and invisible 
+to users.
+
+# Specification
+
+## Protected Code Loader Software Work Flow in Open Enclave
+### Build time encryption 
+    A seperate tool "oeencrypt" is provided at enclave build time to encrypt 
+    the enclave image right before the enclave signing process.  
+### Load time decryption
+    On enclave initialization operation (EINIT operation), right before relocation, 
+    the encrypted enclave image need to be decrypted and than perform the relocation
+    operation.
+
+## Encryption Algorithm in Protected Code Loader
+Using openssl AES-256-GCM as the encryption/decryption algorithm.
+
+## ELF sections Left plaintext
+### PCL table entry 
+A PCL table entry (a section called ".pcltbl" in ELF file)is built into the Enclave 
```

Comment:
If an enclave consists of multiple elf binaries, would each binary contain the PCL sections?

Reply:
while considering the question, I checked dynamic linking branch and the test code, I didn't see any signing performed for the enclave or the dynamic lib linking to the enclave. How did you consider measurement on those multiple modules ? How did you manage heap/stack between those multiple modules? How is oesign used in dynamic linking?

Reply:
@johnvenn Dynamic linking still requires that all shared object files are loaded into the enclave memory space deterministically and measured before initialization, so it doesn't change the enclave measurement model. i.e. instead of loading a single binary, it simply extends that to dynamically loading and binding symbols from dependency binaries as part of the enclave app that is measured prior to initialization. It does not permit runtime loading of new shared objects into the enclave memory space, which would require an extension of the measurement model (such as the use of the experimental EEID feature). oesign behaves as it always does in this regard: it simply takes the enclave measurement, whether consisting of one module or multiple modules, and signs the entire enclave measurement.

The heap/stack model similarly doesn't change at all; those are process-level concepts, not specific to modules loaded for the app, enclave or otherwise. There is some additional C library initialization that needs to be extended when loading multiple modules (e.g. allocation of space for static thread-local variables from all loaded modules per TCS) which you will see extra code to handle in the feature/dynamic_binding branch.